File: Ahorn/entities/SwingingDoor.jl
module SwingingDoor

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/SwingingDoor" SwingingDoor(x::Integer, y::Integer, autoCloseTime::Number=2.0, isDoubleDoor::Bool=true, isLocked::Bool=false, knockbackForce::Integer=150, swingSpeed::Number=3.0)

const placements = Ahorn.PlacementDict(
    "unlocked" => Ahorn.EntityPlacement(SwingingDoor),
    "locked" => Ahorn.EntityPlacement(SwingingDoor)
)

function Ahorn.selection(entity::SwingingDoor)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SwingingDoor, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/swingingDoor.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local swingingDoor = {}

swingingDoor.name = "MaggyHelper/SwingingDoor"
swingingDoor.depth = -50
swingingDoor.fieldInformation = {
    swingSpeed = {
        minimumValue = 0
    },
    knockbackForce = {
        minimumValue = 0
    },
    autoCloseTime = {
        minimumValue = 0
    }
}
swingingDoor.placements = {
    {
        name = "unlocked",
        data = {
            swingSpeed = 3.0,
            knockbackForce = 150,
            isLocked = false,
            isDoubleDoor = true,
            autoCloseTime = 2.0
        }
    },
    {
        name = "locked",
        data = {
            swingSpeed = 3.0,
            knockbackForce = 150,
            isLocked = true,
            isDoubleDoor = true,
            autoCloseTime = 2.0
        }
    }
}

function swingingDoor.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/swinging_door/closed", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    return sprite
end

function swingingDoor.selection(room, entity)
    return utils.rectangle(entity.x - 20, entity.y - 48, 40, 48)
end

return swingingDoor
